src/utils/changes.ts

```diff
@@ -175,3 +181,152 @@ export function prependChangeset(
 
   return markdown.slice(0, startIdx) + newSection + markdown.slice(startIdx);
 }
+
+interface Commit {
+  hash: string;
+  title: string;
+  hasPRinTitle: boolean;
+  pr: string | null;
+  milestone: string | null;
+}
+
+interface Milestone {
+  title?: string;
+  description?: string;
+  state?: string;
+  prs: string[];
+}
+
+// This is set to 8 since GitHub and GitLab prefer that over the default 7 to
+// avoid collisions.
+const SHORT_SHA_LENGTH = 8;
+function formatCommit(commit: Commit): string {
+  let text = `- ${commit.title}`;
+  if (!commit.hasPRinTitle) {
+    const link = commit.pr
+      ? `#${commit.pr}`
+      : commit.hash.slice(0, SHORT_SHA_LENGTH);
+    text = `${text} (${link})`;
+  }
+  return text;
+}
+
+export async function generateChangesetFromGit(
+  git: SimpleGit,
+  rev: string
+): Promise<string> {
+  const { repo, owner } = await getGlobalGithubConfig();
+  const gitCommits = await getChangesSince(git, rev);
+  const commitsQuery = gitCommits
+    .map(({ hash }) => `C${hash}: object(oid: "${hash}") {...PRFragment}`)
+    .join('\n');
+
+  const github = getGithubClient();
+  const { repository: githubCommits } = await github.graphql(`{
+    repository(name: "${repo}", owner: "${owner}") {
+      ${commitsQuery}
+    }
+  }
+
+  fragment PRFragment on Commit {
+    associatedPullRequests(first: 1) {
+      nodes {
+        number
+        milestone {
+          number
+        }
+      }
+    }
+  }`);
+
+  const milestones: {
+    [number: string]: Milestone;
+  } = {};
+  const commits: {
+    [hash: string]: Commit;
+  } = {};
+  const leftovers: Commit[] = [];
+  const missing: Commit[] = [];
+  for (const gitCommit of gitCommits) {
+    const hash = gitCommit.hash;
+
+    const commit = (commits[hash] = {
+      hash: hash,
+      title: gitCommit.message,
+      pr: gitCommit.pr,
+      hasPRinTitle: Boolean(gitCommit.pr),
+      milestone: null,
+    });
+    const commitInfo = githubCommits[`C${hash}`];
+    if (commitInfo) {
+      const pr = commitInfo.associatedPullRequests.nodes[0];
+      if (pr) {
+        commit.pr = pr.number;
+        commit.milestone = pr.milestone?.number;
+      }
+    } else {
+      missing.push(commit);
+    }
+
+    const milestoneNum = commit.milestone;
+    if (milestoneNum != null) {
+      const milestone = milestones[milestoneNum] || { prs: [] };
+      // We _know_ the PR exists as milestones are attached to PRs
```

Comment:
Q: is it possible to get a PR from GitHub that doesn't exist?

Reply:
PR is a concept that GitHub invented, so no. 😁 Even more, the case here is we have a milestone already which can only be associated with a PR or an issue, not a commit directly.

Ideally, we'd have a conditional type here where one says `pr: string | null; milestone: null` and the other part says `pr: string; milestone: string | null` but I got lazy, and also thought this was harder to reason about.

Thoughts?

Reply:
Sounds good! My next question is whether we want to have PRs not attached to any milestone and commits in the same section, but that's more a behavior question and definitely something we can think about later. Just leaving my thoughts.

Reply:
>My next question is whether we want to have PRs not attached to any milestone and commits in the same section, but that's more a behavior question and definitely something we can think about later.

Yup, this was the initial design we agreed on. Change/PR does not matter if they are not tied to a larger milestone for now.